### ISSUE TITLE
Only the first(?) `Screen` associated with an app gets the title/sub-title in any mounted `Header`

### ISSUE THREAD

**Issue by davep** [Textualize/textual#1307]
Consider this code:

```python
from textual.app import App, ComposeResult
from textual.screen import Screen
from textual.widgets import Header, Footer, Label

class First( Screen ):

    BINDINGS = [
        ( "n", "switch_screen( 'second' )", "Second Screen" )
    ]

    def compose( self ) -> ComposeResult:
        yield Header()
        yield Label(
            "This is the first screen\n"
            "Notice the app title in the header."
        )
        yield Footer()

class Second( Screen ):

    BINDINGS = [
        ( "n", "switch_screen( 'first' )", "First Screen" )
    ]

    def compose( self ) -> ComposeResult:
        yield Header()
        yield Label(
            "This is the second screen\n"
            "Notice the total lack of app title in the header."
        )
        yield Footer()

class MultiScreen( App[ None ] ):

    CSS = """
    Screen {
        align: center middle;
    }
    Label {
        width: auto;
    }
    """
    TITLE = "Very Cool App"

    SUB_TITLE = "These titles show off how proud I am about it"

    SCREENS = {
        "first": First,
        "second": Second
    }

    def on_compose( self ) -> None:
        self.push_screen( "first" )

if __name__ == "__main__":
    MultiScreen().run()
```

When run, only the `First` screen shows the app title and subtitle in its mounted `Header`, `Second` doesn't.

**Comment by davep** [comment]
Interesting. That feels like a good first place to look.

**Comment by dehesselle** [comment]
I tried to debug this. I've added logging to `on_mount` in `_header.py`

```python
def on_mount(self) -> None:
    def set_title(title: str) -> None:
        # let' see what's going on
        logging.debug(f"Header set_title {title} for {self.parent.parent}")
        self.query_one(HeaderTitle).text = title

    def set_sub_title(sub_title: str) -> None:
        self.query_one(HeaderTitle).sub_text = sub_title
```

and compose my screens like this

```python
def compose(self) -> ComposeResult:
    yield Container(
        Header(),
        Footer(),
        # ...other stuff here...
    )
```

For my first screen named `WelcomeScreen` it shows me
```text
DEBUG:root:Header set_title MyApp for WelcomeScreen()
```
For my second screen named `CredentialsScreen` it shows me
```text
DEBUG:root:Header set_title MyApp for WelcomeScreen(pseudo_classes={'focus-within'})
```
instead of
```text
DEBUG:root:Header set_title MyApp for CredentialsScreen()
```
So this explains why only the first screen works, but I don't understand what's causing this.

**Comment by nitzan-shaked** [comment]
If I had to guess, I'd say the issue is the code in `mount()`: if registers `set_title` and `set_sub_title` as watchers, but they are THE SAME for both instances of Header() (one for each screen). The reason is that they don't take `self`. One possible fix is to make the instance-methods and register `self.set_title` etc.

(Notice the code in reactive.py does not register a "callback" for the second time).

**Comment by dehesselle** [comment]
>If I had to guess, I'd say the issue is the code in mount(): if registers set_title and set_sub_title as watchers, but they are THE SAME for both instances of Header() (one for each screen). The reason is that they don't take self. One possible fix is to make the instance-methods and register self.set_title etc.

That is actually the first thing I've tried, but it didn't help. Maybe it's a combination of this and something else... (unless I made mistake testing this).

**Comment by nitzan-shaked** [comment]
Found it. The problem is line 252 of reactive.py, the one that says `for watcher in watchers`. The problem is that `invoke_watcher()` is only actually called IF we still don't think that a compute is required. Once a compute is required, the rest of the watchers are not invoked. (notice the OR). Changing it to something such as:

```
        for watcher in watchers:
            x = invoke_watcher(watcher, old_value, value)
            require_compute = require_compute or x
```

... solves the issue.

This is a hack, or at least I don't know that it isn't. But this should be enough for the team to fix this, I assume.

**Comment by nitzan-shaked** [comment]
To be clear, the real solve requires a bit more to account for coroutines etc. Maybe. All I'm saying is that "it's there".

**Comment by nitzan-shaked** [comment]
(The design seems to think this: the first watcher which is a coro will return False from invoke_watcher, and the rest of the watchers won't be called. But that's not bad: when the coro finishes and await_watcher gets invoked, it will call _compute. And _compute will invoke setattr, and that will re-trigger, and the other watchers will get a chance to run.

BUT: there is a flaw in this logic; if _compute returns the same value as the current value, then the __set__ descriptor will not re-trigger anything (unless always_update is set, but that's not always the case). so the other watchers don't get a chance to run.

So: if you have a bunch of watchers, and the some of them are coros, and the resulting _compute returns the same value as current value, the rest don't get to run. At least that's what I make of browsing the code in a terminal now. I could be wrong.

I'm the maintainers can come up with better solutions than what I might suggest here, so...)

**Comment by nitzan-shaked** [comment]
@davep bumping this, just wanting to make this gets attention. I believe (?) I found the cause (see above), and _if_ I'm correct then this will affect other things all around. I could very well be wrong though, so was hoping someone could have a look.

**Comment by davep** [comment]
@nitzan-shaked Along with plenty of other things it's in the queue to be looked at.

**Comment by rodrigogiraoserrao** [comment]
Your explorations can't be too far from the truth. Merging the draft PR #1539 at commit c348a1f seems to fix this issue, so I'll wait for 1539 resolution before coming back to this.

**Comment by rodrigogiraoserrao** [comment]
It ended up not being PR 1539 but this is now fixed thanks to #1558
